posthog/auth.py

```diff
@@ -68,27 +69,37 @@ def find_key(
         return key_with_source[0] if key_with_source is not None else None
 
     @classmethod
+    @transaction.atomic
     def validate_key(cls, personal_api_key_with_source):
         from posthog.models import PersonalAPIKey
 
         personal_api_key, source = personal_api_key_with_source
         personal_api_key_object = None
+        mode_used = None
 
-        for iterations in PERSONAL_API_KEY_ITERATIONS_TO_TRY:
-            secure_value = hash_key_value(personal_api_key, iterations=iterations)
+        for mode, iterations in PERSONAL_API_KEY_MODES_TO_TRY:
+            secure_value = hash_key_value(personal_api_key, mode=mode, iterations=iterations)
             try:
                 personal_api_key_object = (
                     PersonalAPIKey.objects.select_related("user")
                     .filter(user__is_active=True)
                     .get(secure_value=secure_value)
                 )
+                mode_used = mode
                 break
             except PersonalAPIKey.DoesNotExist:
                 pass
 
         if not personal_api_key_object:
             raise AuthenticationFailed(detail=f"Personal API key found in request {source} is invalid.")
 
+        # Upgrade the key if it's not in the latest mode. We can do this since above we've already checked
+        # that the key is valid in some mode, and we do check for all modes one by one.
+        if mode_used != PERSONAL_API_KEY_MODES_TO_TRY[0][0]:
```

Comment:
NIT: Could we not just be explicit here and say? Feels a bit weird to be accessing it this way
```suggestion
        if mode_used != "sha256":
```

Reply:
You mean in case anyone ever decides to make something else the first to try? Then it would auto-upgrade 😅 
Yeah maybe be explicit here.